overhaul of readme to make more consistent and read-able (#98)
dbt-labs/dbt-audit-helper: README.md

Useful macros when performing data audits

# Contents
- [Installation instructions](#installation-instructions)
- [Compare Data Outputs](#compare-data-outputs)
  - [compare\_relations (source)](#compare_relations-source)
  - [compare\_queries (source)](#compare_queries-source)
  - [compare\_row\_counts (source)](#-compare_row_counts-source)
- [Compare Columns](#compare-columns)
  - [compare\_column\_values (source)](#compare_column_values-source)
  - [compare\_all\_columns (source)](#compare_all_columns-source)
  - [compare\_which\_columns\_differ (source)](#compare_which_columns_differ-source)
  - [compare\_relation\_columns (source)](#compare_relation_columns-source)
- [Print Output To Logs](#print-output-to-logs)
- [Use Output For Custom Singular Test](#use-output-for-custom-singular-test)

# Installation instructions
New to dbt packages? Read more about them [here](https://docs.getdbt.com/docs/building-a-dbt-project/package-management/).
1. Include this package in your `packages.yml` file — check [here](https://hub.getdbt.com/dbt-labs/audit_helper/latest/) for the latest version number.
2. Run `dbt deps` to install the package.

# Compare Data Outputs

> ## compare_relations ([source](macros/compare_relations.sql))

Reviewer: Do you think we should treat compare_relations as a wrapper around compare_queries, and reorder/rephrase them with that in mind? So lean into documentation for compare_queries and then say "Also for convenience compare_relations exists assuming column compatibility etc"

Author: That makes sense to me!

> This macro generates SQL that can be used to do a row-by-row comparison of two relations. This macro is particularly useful when you want to check that a refactored model, or a model that you are moving over from a legacy system, match up.

Reviewer: ```suggestion
This macro generates SQL that can be used to do a row-by-row comparison of two relations. This macro is particularly useful when you want to check that a refactored model (or a model that you are moving over from a legacy system) are identical.

Each relation must have the same columns with the same names, but they do not have to be in the same order. Use `exclude_columns` if some columns only exist in one relation. 
```

{% endset %}

{{ audit_helper.compare_queries(
    a_query = old_query,
    b_query = new_query,
    primary_key = "order_id"
) }}

```

## compare_row_counts ([source](macros/compare_row_counts.sql))
This macro does a simple comparison of the row counts in two relations. 

### Output:
Calling this macro on two different relations will return a very simple table comparing the row counts in each relation. 

| relation_name                                | total_records  |
|----------------------------------------------|---------------:|
| target_database.target_schema.my_a_relation  |     34,231     |
| target_database.target_schema.my_b_relation  |     24,789     |

### Usage:
```sql

{% set old_relation = adapter.get_relation(
      database = "old_database",
      schema = "old_schema",
      identifier = "fct_orders"
) -%}

{% set dbt_relation = ref('fct_orders') %}

{{ audit_helper.compare_row_counts(
    a_relation = old_relation,
    b_relation = dbt_relation
) }}

```

### Arguments:
* `a_relation` and `b_relation`: The [relations](https://docs.getdbt.com/reference/dbt-classes#relation) you want to compare. 

# Compare Columns 

> ## compare_column_values ([source](macros/compare_column_values.sql))

Reviewer: When would you ever want this query instead of compare_all_columns? Should we remove it altogether? It requires you to already know which column your problem is in, instead of helping you diagnose where the problem is. 

At minimum we probably want to swap them around and deemphasise it, but I sorta feel like we could remove it altogether.

Author: I disagree - I think this is a super valuable one! It does require you to know which column your problem is in (you could use the new `compare_which_columns_differ` for that, but it's way less computationally expensive to look at a single column vs. all of the columns!

Author: There is `compare_column_values_verbose` which we don't currently document, should we?

> This macro generates SQL that can be used to compare a column's values across two queries. This macro is useful when you've used the `compare_queries` macro (above) and found that a significant number of your records don't match; So now you want to find the discrepancies of 1 specific column.

Reviewer: ```suggestion
This macro generates SQL that can be used to compare a column's values across two queries. This macro is useful when you've used the `compare_queries` macro (above) and found that a significant number of your records don't match and want to understand how many discrepancies are caused by a single column.
```

### Output:
The generated query returns a summary of the count of rows where the column's values:
- match perfectly 
- differ
- are null in `a` or `b` or both
- are missing from `a` or `b`

| match_status                | count  | percent_of_total |
|-----------------------------|-------:|-----------------:|
| ✅: perfect match            | 37,721 | 79.03            |
| ✅: both are null            | 5,789  | 12.13            |
| 🤷: missing from a          | 5     | 0.01             |
| 🤷: missing from b          | 20     | 0.04             |
| 🤷: value is null in a only | 59     | 0.12             |
| 🤷: value is null in b only | 73     | 0.15             |

> | 🙅: ‍values do not match    | 4,064  | 8.51             |

Reviewer: ```suggestion
| ❌: ‍values do not match    | 4,064  | 8.51             |
```

{% endset %}

{% set new_query %}
    select * from {{ ref('dim_product') }}
{% endset %}

{{ audit_helper.compare_column_values(
    a_query = old_query,
    b_query = new_query,
    primary_key = "product_id",
    column_to_compare = "status"
) }}

```

## compare_all_columns ([source](macros/compare_all_columns.sql))
Similar to `compare_column_values`, except it can be used to compare _all_ columns' values across two _relations_. 

### Output:
By default, the generated query returns a summary of the count of rows where the each column's values:
- match perfectly 
- differ
- are null in `a` or `b` or both
- are missing from `a` or `b`

| column_name  | perfect_match  | null_in_a | null_in_b | missing_from_a | missing_from_b | conflicting_values |
|-------|-------:|------:|-----------------:|------:|------:|------:|
| order_id  | 10 | 0 | 0 | 0 | 0 | 0 |
| order_date  | 2 | 0 | 0 | 0 | 0 | 8 |
| order_status | 6 | 4 | 4 | 0 | 0 | 0 |

Setting the `summarize` argument to `false` lets you check the match status of a specific column value of a specifc row:

| primary_key | column_name | perfect_match  | null_in_a | null_in_b | missing_from_a | missing_from_b | conflicting_values |

> |-------||-------|-------:|------:|-----------------:|------:|------:|------:|

Reviewer: ```suggestion
|-------|-------|-------:|------:|-----------------:|------:|------:|------:|
```

| status      | 4                  | 5                  | character varying | character varying |                      False |                True |     False |     False |    True |
| amount      | 5                  | 4                  | bigint            | bigint            |                      False |                True |     False |     False |    True |

Note: For adapters other than BigQuery, Postgres, Redshift, and Snowflake, the ordinal position is inferred based on the response from dbt Core's `adapter.get_columns_in_relation()`, as opposed to being loaded from the information schema.

### Arguments:
* `a_relation` and `b_relation`: The [relations](https://docs.getdbt.com/reference/dbt-classes#relation) you want to compare.

### Usage:
```sql

{% set old_relation = adapter.get_relation(
      database = "old_database",
      schema = "old_schema",
      identifier = "fct_orders"
) -%}

{% set dbt_relation = ref('fct_orders') %}

{{ audit_helper.compare_relation_columns(
    a_relation=old_relation,
    b_relation=dbt_relation
) }}

```

> # Print Output To Logs

Reviewer: ```suggestion
## Advanced Usage 
### Print Output To Logs
```

Author: I'm all for this, but we should add to the table of contents as well!

{% endmacro %}
```

To run the macro, execute `dbt run-operation print_audit_output()` in the command bar.

> # Use Output For Custom Singular Test

Reviewer: ```suggestion
### Use Output For Custom Singular Test
```
